feat(rome_js_analyze): lint/correctness/noDupeKeys

jeysal opened the pull request:
## Summary

Resolves #3364 

Note that I've stopped short of feature parity with ESLint because the PR already got pretty big. The rule can be extended later. I can open a follow up to #3364 when this is merged.
The cases that ESLint covers but this doesn't can be seen in the test file - number-keyed properties and computed-string-literal-keyed properties.

## Test Plan

Used standard lint test infrastructure, had to extend slightly for the suggested fix action

jeysal commented on crates/rome_js_analyze/tests/specs/nursery/noDupeKeys.js:
```diff
@@ -0,0 +1,36 @@
+// invalid
```
I'm hitting a strange parse issue with this test file as soon as I add seemingly any one of the invalid cases:
```
There should be no errors in the file "/home/seckinger/proj/rometools/crates/rome_js_analyze/tests/specs/nursery/noDupeKeys.js" but the following errors where 
present:
```
(with no actual errors printed, just the `Parsed tree:` follows)

Assuming this also happens on CI, could someone take a look?

jeysal replied:
Oh, so this is because of the code actions I added which (as I know, see other comment), aren't fully working yet. I didn't even realize these are being run in the tests, was actually wondering whether those are not tested at all as it seemed to me

jeysal commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,254 @@
+    }
+}
+
+fn get_property_name(member: &JsAnyObjectMember) -> Option<String> {
```
Not sure if this should be some trait in the syntax ext?

jeysal commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,254 @@
+        PropertyConflict(_, PropertyDefinition(property_type, node)): &Self::State,
+    ) -> Option<JsRuleAction> {
+        let mut batch = ctx.root().begin();
+        batch.remove_node(node.clone());
```
This has an issue: When applying the suggestion, the node is removed but not the comma trailing it. Because of this, applying the suggestion fails. This seems like a bug in the removal impl.

95th replied:
Should there be an action on this at all? Eslint doesn't show fix.

jeysal replied:
I don't see a reason not to, it's by far the most likely user action to delete the property I would say so someone might find it useful. As long as we don't pretend it's a safe fix. We could optionally limit it further and only offer the fix e.g. if the right side is a literal so that it's safe.

jeysal replied:
From what I understand, this could be intentional if the comma is considered trivia, and in that case I'll have to manipulate that trivia manually?

leops replied:
Separated lists are represented as a sequence of AST nodes and separator tokens, so the comma is actually a separate element of the list for the node itself. This means in order to remove the trailing comma for a node in the list, you also need to issue a `remove_token` operation for the separator token to the batch mutation

jeysal commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,254 @@
+                    diagnostic.detail(setter_node.range(), "Overwritten by this setter.")
+                }
+                PropertyType::Value => {
+                    let diagnostic =
```
TODO: This is awkward. We generally only print a hint for the last of the nodes overwriting a property and not for all nodes following it that overwrite it. But in case of a property value overwritten by both a getter and a setter, we hint for both of them. We should probably either only hint for whichever is later in this case too, or always hint for all overwriting nodes. This will likely also simplify the data structures.

jeysal commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,254 @@
+        for member in node
+            .members()
+            .into_iter()
+            // Note that we iterate from last to first property, so that we highlight properties being overwritten as problems and not those that take effect.
```
This might be a bit of a hot take, and ESLint does it the other way.
But I think it's the right way, esp. considering the fix suggestion.

jeysal commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
````diff
@@ -0,0 +1,254 @@
+    /// ```
+    ///
+    pub(crate) NoDupeKeys {
+        version: "10.0.0",
````
Marker: Change this if we don't make it in time

jeysal commented on crates/rome_diagnostics_categories/src/categories.rs:
```diff
@@ -75,14 +75,15 @@ define_dategories! {
 
     // nursery
-    "lint/nursery/useFlatMap": "https://rome.tools/docs/lint/rules/useFlatMap",
+    "lint/nursery/noBannedTypes":"https://rome.tools/docs/lint/rules/noBannedTypes",
```
Went ahead and pressed sort on this

jeysal reviewed (commented):
There are some things to figure out still, but overall it turned out pretty clean I'd say.
Not sure yet if I'll continue on the weekend or later.
I hit some infra issues that would probably be quickest solved with some help from someone who's already had to deal with similar things, see my comments.

jeysal commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,254 @@
+                        diagnostic.detail(getter_node.range(), "Overwritten by this getter.");
+                    let diagnostic =
+                        diagnostic.detail(setter_node.range(), "Overwritten by this setter.");
+                    diagnostic
```
This gets a lint warning for returning a let binding. Pretty annoying warning tbh as it wants you to break the symmetry of the `.detail()` calls here 🙄
If I address the comment above it'll probably be a non-issue anyway as this code will change

jeysal commented on crates/rome_js_analyze/src/utils/tests.rs:
```diff
@@ -43,28 +49,55 @@ pub fn assert_rename_nok(before: &str) {
+/// Search an identifier named "a" and remove the entire node of type Anc around it.
 /// Asserts the removal worked.
-pub fn assert_remove_ok(before: &str, expected: &str) {
+pub fn assert_remove_identifier_a_ok<Anc: AstNode<Language = JsLanguage> + Debug>(
```
I had to generalize this test util a bit so that it can look for any identifier token and then you tell it how much around the token you'd like to remove by giving it an ancestor type to look for

jeysal commented on crates/rome_js_analyze/src/utils/tests.rs:
```diff
@@ -113,8 +146,7 @@ pub fn ok_find_attributes_by_name() {
         .descendants()
-        .filter_map(rome_js_syntax::JsxAttributeList::cast)
-        .next()
+        .find_map(rome_js_syntax::JsxAttributeList::cast)
```
autofix :sweat_smile:

jeysal commented on crates/rome_js_analyze/tests/specs/nursery/noDupeKeys.js:
```diff
@@ -0,0 +1,43 @@
+
+// valid for now
+
+// ESLint already catches properties keyed with different-formatted number literals, we haven't implemented it yet.
```
This PR has grown pretty big already so I'll leave this for a future extension to the rule :sweat_smile:

jeysal commented on crates/rome_diagnostics_categories/src/categories.rs:
```diff
@@ -75,14 +75,15 @@ define_dategories! {
 
     // nursery
-    "lint/nursery/useFlatMap": "https://docs.rome.tools/lint/rules/useFlatMap",
+    "lint/nursery/noBannedTypes":"https://docs.rome.tools/lint/rules/noBannedTypes",
```
went ahead and pressed sort on these as there was no apparent order to them

jeysal commented:
This PR is ready for review now! :)
I left some comments with explanations and some with minor implementation questions, but overall I'm pretty happy, esp. with the diagnostic as well.
As noted in the PR description, this rule is not at feature parity with the ESLint one yet. It was already getting pretty big, so this can be shipped and then extended later.

MichaReiser commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
````diff
@@ -0,0 +1,262 @@
+    /// ```
+    ///
+    pub(crate) NoDupeKeys {
+        version: "10.1.0",
````
```suggestion
        version: "11.0.0",
```

There are some breaking changes ahead of us :D

MichaReiser commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,262 @@
+
+#[derive(Clone)]
+enum DefinedProperty {
+    Getter(JsAnyObjectMember),
+    Setter(JsAnyObjectMember),
```
Would it be possible to narrow the types to `JsGetterObjectMember` and `JsSetterObjectMember`?

jeysal replied:
Checked it right now and pretty much immediately noticed a problem - for `Value` we would need a union of `JsPropertyObjectMember`, `JsShorthandPropertyObjectMember`, and `JsMethodObjectMember`, and thus extend the data structure. Or we would have a weirdly asymmetric `Getter(JsGetOM), Setter(JsSetOM), Value(JsAnyOM)`.

On the other hand, we currently have nothing to gain from narrowing. We don't later cast back to specific object member types. In fact, the only reason I changed from just storing the `TextRange` in this data structure to storing the entire node is to pass the node on to a batch mutation function in the action.

For these two reasons I think we should leave it as is as it would only introduce extra complexity.

MichaReiser replied:
The main advantage I'm seeing of narrowing the type for Getters/setters is that it more clearly communicates the intent to readers. 

We can achieve the same for `GetterSetter` and `Value` by documenting the variants the value can hold.

MichaReiser commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,262 @@
+    Getter(JsAnyObjectMember),
+    Setter(JsAnyObjectMember),
+    GetterSetter(JsAnyObjectMember, JsAnyObjectMember),
+    Value(JsAnyObjectMember),
```
Is it correct that the last variant covers methods and properties? Renaming the variant to `PropertyOrMethod` may be more explicit if that's the case.

jeysal replied:
I was thinking of this in terms of object descriptors, which have either `get` and `set` or a `value`, with methods just being values as well.
I don't feel strongly about keeping this wording, but just to explain where my inspiration for the terminology came from, perhaps knowing that you might find it clean as well.

`PropertyOrMethod` as a specific counter proposal is problematic, because I've been using "property" to refer to object properties in general, whether get-set or value.

My recommendation would be to keep it. We could optionally change `Getter` to `Get` and `Setter` to `Set` to even better resemble object property descriptors.
If we end up changing this, reminder to change it in the diagnostic messages as well, those also use this wording.

MichaReiser replied:
Rome's internal naming convention to this point is that we call object "properties" members and only use the term "property" for "value" properties. 

I don't think there's an established standard on how we name "properties" in diagnostics, going with properties. I think properties can make more sense as most people are used to it. Tough, I had never heard of a value property before (in my mind: Methods, setters, getters, spread, and properties are the rest).

MichaReiser commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,262 @@
+enum PropertyType {
+    Getter,
+    Setter,
+    Value,
```
Is it correct that the last variant covers methods and properties? Renaming the variant to PropertyOrMethod may be more explicit if that's the case.

jeysal replied:
Replied on the other similar comment, let's discuss there

MichaReiser commented on crates/rome_js_analyze/src/analyzers/nursery/no_dupe_keys.rs:
```diff
@@ -0,0 +1,262 @@
+            .into_iter()
+            // Note that we iterate from last to first property, so that we highlight properties being overwritten as problems and not those that take effect.
+            .rev()
+            .filter_map(|result| result.ok())
```
Nit: Is this the same as calling `flatten()`?

jeysal replied:
I think so. I find it a bit less clear, but probably only because I haven't worked with Rome's syntax structures as much yet. I'll change it

MichaReiser commented on crates/rome_js_analyze/src/utils/batch.rs:
```diff
@@ -168,14 +173,36 @@ impl JsBatchMutation for BatchMutation<JsLanguage> {
+                for element in parent.elements() {
+                    if element.node() == Ok(member) {
+                        self.remove_node(member.clone());
+                        if let Some(comma) = element.trailing_separator().ok().flatten() {
```
Nit:

```suggestion
                        if let Some(Ok(comma)) = element.trailing_separator() {
```

jeysal replied:
Yup, does look cleaner. I copied this from some other batch removal util we already had, I'll change it in all of them

MichaReiser reviewed (approved):
Well done. I have a few small nits, and we must wait to merge until the 10.0.1 patch release is out of the door.

jeysal commented:
All comments addressed with changes or explanations why they were left unchanged.
Also merged in main.
Should be ready to go once 11.0.0 merges are starting!
Ping me if it becomes out of date with main or so.

jeysal commented:
@MichaReiser I refactored quite a lot and should have addressed your concerns about too wide typings and about blurry naming conventions.

Member definitions now clearly use the existing terminology that you mentioned.
Defined properties now make it clear that they refer to property descriptors that we think will make it into the final object.

I also used more impls and traits to perhaps make the core algorithm of the rule itself easier to read.
It's overall a couple tens of lines plus in the diff, but perhaps you'll find it clearer.
Maybe the type narrowing will be of benefit in future extensions to the rule.

MichaReiser reviewed (commented):
We're good to go after you resolved the merge conflicts.
